src/Shared.TextureContainer/Shared.TextureContainerManager.cs

```diff
@@ -0,0 +1,227 @@
+            }
+
+            public byte[] Data => key.data;
+
+            public Texture Texture
+            {
+                get
+                {
+                    if (_texture == null && key.data != null)
+                        _texture = TextureFromBytes(key.data, key.format, key.mipmaps);
+
+                    return _texture;
+                }
+            }
+            
+            public void Destroy()
+            {
+                if (_texture != null)
+                {
+                    UnityEngine.Object.Destroy(_texture);
+                    _texture = null;
+                }   
+            }
+        }
+
+        /// <summary>
+        /// Acquire TextureHolder. If it already exists, return it.
+        /// </summary>
+        /// <param name="filePath"></param>
+        /// <returns></returns>
+        public static Token Acquire( string filePath )
+        {
+            return Acquire(LoadTextureBytes(filePath));
+        }
+
+        /// <summary>
+        /// Acquire TextureHolder. If it already exists, return it.
+        /// </summary>
+        /// <param name="texBytes"></param>
+        /// <returns></returns>
+        public static Token Acquire( byte[] texBytes )
+        {
+            if (texBytes == null)
+                throw new ArgumentNullException(nameof(texBytes));
+
+            TextureKey key = new TextureKey(texBytes);
+            if (!_textureHolder.TryGetValue(key, out var holder))
+                holder = _textureHolder[key] = new Token(key);
+
+            ++holder.refCount;
+            return holder;
+        }
+
+        /// <summary>
+        /// Release the TextureHolder. 
+        /// If there are zero TextureHolders with the same texture, the texture is released.
+        /// </summary>
+        /// <param name="holder"></param>
+        public static void Release( Token holder )
+        {
+            if (--holder.refCount > 0)
+                return;
+
+            holder.Destroy();
+            _textureHolder.Remove(holder.key);
+        }
+
+        /// <summary>
+        /// Convert a byte array to Texture2D.
+        /// </summary>
+        /// <param name="texBytes">Byte array containing the image</param>
+        /// <param name="format">TextureFormat</param>
+        /// <param name="mipmaps">Whether to generate mipmaps</param>
+        /// <returns></returns>
+        private static Texture TextureFromBytes(byte[] texBytes, TextureFormat format, bool mipmaps)
+        {
+            if (texBytes == null || texBytes.Length == 0) return null;
+
+            //LoadImage automatically resizes the texture so the texture size doesn't matter here
+            Texture2D tex = new Texture2D(2, 2, format, mipmaps);
+
+            try
+            {
+                tex.LoadImage(texBytes);
+
+                RenderTexture rt = new RenderTexture(tex.width, tex.height, 0);
+                rt.useMipMap = mipmaps;
+                Graphics.Blit(tex, rt);

```

Comment:
Is copying the texture into a new RenderTexture necessary, instead of simply returning `tex`?

The summary description of the method states that Texture2D is returned.

Reply:
The summary has been revised.

I would rather hear why from me.
Wasn't this a deliberate process?

I made the following changes locally.
```
Texture2D tex = new Texture2D(2, 2, format, mipmaps);
tex.LoadImage(texBytes);
return tex;
```

Loading times were virtually unchanged.
I looked at some scenes and I don't think there was any change in appearance.
For reasons unknown, memory usage after loading increased by about 1.5 GB at most.

I imagine that RenderTexture only has data in GPU memory, but Texture2D has textures in both regular memory and GPU memory, so the memory usage would be inflated.


The code comes from this commit.
https://github.com/IllusionMods/KK_Plugins/commit/50a5c7f7f06bbee41d278db24894c996492820dd

To speed up normal map changes ......?
There was no difference in load times in my environment, but that may be because I didn't have any characters to change normal maps.

Reply:
You are right, RenderTexture exists only on the GPU, while Texture2D has a copy in main memory.

If nothing needs a Texture2D then this is better. Maybe also add a comment explaining why RenderTexture is used?

Reply:
Added comment.